lib/sycamore/sycamore/connectors/common.py

```diff
@@ -88,49 +112,86 @@ def flatten_data(
     return items
 
 
-def unflatten_data(data: dict[str, Any], separator: str = ".") -> dict[Any, Any]:
-    result: dict[Any, Any] = {}
+def unflatten_data(data: dict[Any, Any], separator: str = ".") -> dict[Any, Any]:
+    """
+    Unflattens a dictionary with keys that contain separators into a nested dictionary. The separator can be escaped,
+    and if there are integer keys in the path, the result will be a list instead of a dictionary.
+    """
 
-    def parse_key(key: str) -> list:
-        # Handle escaped separator
+    def split_key(key: str, separator: str = ".") -> list[str]:
+        """
+        Splits the key by separator (which can be multiple characters), respecting escaped separators.
+        """
         parts = []
         current = ""
-        escape = False
-        for char in key:
-            if escape:
-                if char == separator:
-                    current += separator
+        i = 0
+        while i < len(key):
+            if key[i] == "\\":
+                # Escape character
+                if i + 1 < len(key):
+                    current += key[i + 1]
+                    i += 2
                 else:
-                    current += "\\" + char
-                escape = False
-            elif char == "\\":
-                escape = True
-            elif char == separator:
+                    # Trailing backslash, treat it as literal backslash
+                    current += "\\"
+                    i += 1
+            elif i + len(separator) < len(key) and key[i : i + len(separator)] == separator:
```

Comment:
pretty sure python just truncates if you ask for a slice that's too big. ie
[1, 2, 3, 4, 5][3:7] == [4, 5]

Reply:
Changed.